packages/server/src/core/middleware.ts

```diff
@@ -74,29 +139,32 @@ export type MiddlewareFunction<
     };
   }): Promise<MiddlewareResult<TParamsAfter>>;
   _type?: string | undefined;
+  piped?: Array<MiddlewareFunction<TParams, TParamsAfter>>;
 };
 
-/**
- * @internal
- */
-// FIXME this should use RootConfig
 export function createMiddlewareFactory<TConfig extends AnyRootConfig>() {
-  return function createMiddleware<TNewParams extends ProcedureParams>(
-    fn: MiddlewareFunction<
-      {
-        _config: TConfig;
-        _ctx_out: TConfig['$types']['ctx'];
-        _input_out: unknown;
-        _input_in: unknown;
-        _output_in: unknown;
-        _output_out: unknown;
-        _meta: TConfig['$types']['meta'];
-      },
-      TNewParams
-    >,
+  function createNewBuilder(
+    previousMiddleware: MiddlewareFunction<any, any>,
+    newMiddleware: MiddlewareFunction<any, any>,
   ) {
-    return fn;
-  };
+    newMiddleware.piped = [
+      ...(previousMiddleware.piped ?? [previousMiddleware]),
+      newMiddleware,
+    ];
```

Comment:
Nononono

Reply:
I know it’s bad 😅

I kept bouncing between variations of this and procedurebuilder